ze/btx_zeinterval_callbacks.cpp

```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
+    ze_device_handle_t hDevice, zes_freq_handle_t hFrequency, uint32_t domainIdx,
+    size_t _pFreqState_val_length, zes_freq_state_t *pFreqState_val) {
+  DeviceHash uuid_idx = get_device_hash(usr_data, hostname, vpid, hDevice);
+  btx_push_message_sampling_frequency(btx_handle, hostname, ts, uuid_idx.hash, uuid_idx.deviceIdx,
+                                      (uint64_t)hFrequency, domainIdx, pFreqState_val->actual);
 }
 
-static void lttng_ust_ze_sampling_copyEngine_callback(void *btx_handle, void *usr_data, int64_t ts,
-                                                      const char *hostname, int64_t vpid,
-                                                      uint64_t vtid, ze_device_handle_t hDevice,
-                                                      uint32_t subDevice, uint64_t activeTime,
-                                                      uint64_t sampling_ts) {
+// Properties
+static void lttng_ust_ze_sampling_deviceProperties_callback(
+    void *btx_handle, void *usr_data, int64_t ts, const char *hostname, int64_t vpid, uint64_t vtid,
+    zes_device_handle_t hDevice, uint32_t deviceIdx, size_t _pDeviceProperties_val_length,
+    zes_device_properties_t *pDeviceProperties_val) {
   auto *data = static_cast<data_t *>(usr_data);
-  auto [it, inserted] = data->device_copyEngine_ref.insert(
-      {{hostname, vpid, hDevice, subDevice}, {activeTime, sampling_ts, ts}});
-  // First entry
-  if (inserted)
-    return;
-  auto &[prev_activeTime, prev_sampling_ts, prev_ts] = it->second;
+  data->sampling_device_property[{hostname, vpid, hDevice}] = {*pDeviceProperties_val, deviceIdx};
+}
 
-  /* Per doc: When taking the delta, the difference between timestamp samples
-   * could be 0, if the frequency of sampling the snapshots is higher than the
-   * frequency of the timestamp update. */
-  if (prev_sampling_ts == sampling_ts)
-    return;
+static void lttng_ust_ze_sampling_subDeviceProperties_callback(
+    void *btx_handle, void *usr_data, int64_t ts, const char *hostname, int64_t vpid, uint64_t vtid,
+    zes_device_handle_t hDevice, ze_device_handle_t hSubDevice,
+    size_t _pSubDeviceProperties_val_length, ze_device_properties_t *pSubDeviceProperties_val) {
+  auto *data = static_cast<data_t *>(usr_data);
+  data->sampling_sub_device_property[{hostname, vpid, (ze_device_handle_t)hSubDevice}] =
+      *pSubDeviceProperties_val;
+}
+
+static void lttng_ust_ze_sampling_fabricPortProperties_callback(
+    void *btx_handle, void *usr_data, int64_t ts, const char *hostname, int64_t vpid, uint64_t vtid,
+    ze_device_handle_t hDevice, zes_fabric_port_handle_t hFabricPort,
+    size_t _pFabricPortProperties_val_length,
+    zes_fabric_port_properties_t *pFabricPortProperties_val) {
+  auto *data = static_cast<data_t *>(usr_data);
+  data->fabricPort_property[{hostname, vpid, (ze_device_handle_t)hDevice,
+                             (zes_fabric_port_handle_t)hFabricPort}] = *pFabricPortProperties_val;
```

Comment:
Why the cast? `hFabricPort` is already a `zes_fabric_port_handle_t`

Reply:
Removed, thanks.